### PR TITLE
More friendly experience when presenting repack options

### PR DESCRIPTION
**Pull request by ChristoferMendes**
## CHANGES MADE

- Add a new file for getting info from title (multiplayer/multixx)
- Add new components (Tag, Tooltip)
- Add fetch of peers/seeders (with 2 minutes of cache)
- Add language of installer ('ru' for Xatab, 'en' for others)

<details>
<summary>UI</summary>

repacks-modal
</details>

#225

### CODE REVIEW

**Comment by JackEnx** [comment]
run "yarn lint", there some errors to fix.

image

File: src/main/services/open-web-torrent.ts [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+      const response = await axiosRef.get(endpoint, { params });
+      const { seeds, peers } = this.transformResponseError(response);
+
+      return {
+        seeders: seeds,
+        peers: peers,
+      };
+    } catch (e) {
+      const error = e as Error;
+      throw new Error(error.message);
```
Comment (JackEnx) [inline src/main/services/open-web-torrent.ts:47]: doesn't this have the same effect as not using try/catch?

File: src/main/services/repack-tracker/helpers.ts [hunk 2]
```diff
@@ -5,7 +5,7 @@ import { repackRepository } from "@main/repository";
 
 import { QueryDeepPartialEntity } from "typeorm/query-builder/QueryPartialEntity";
 
-export const savePage = async (repacks: QueryDeepPartialEntity<Repack>[]) =>
+export const saveRepacks = async (repacks: QueryDeepPartialEntity<Repack>[]) =>
```
Comment (JackEnx) [inline src/main/services/repack-tracker/helpers.ts:8]: is there any specific reason for changing the name of function in this pr?

Comment (ChristoferMendes) [reply src/main/services/repack-tracker/helpers.ts:8]: Yeah, sorry, I didn't specify this. But my goal with this change is to improve readability, as the function does not save any page, but instead saves the repacks on the database. While I was reading the code I found this suitable. I will enhance the description to contain this

File: src/main/services/open-web-torrent.ts (hunk 1 above)
Comment (ChristoferMendes) [reply src/main/services/open-web-torrent.ts]: Yes, correct. Updated

**Comment by ChristoferMendes** [comment]
@JackEnx, could you revisit this? Updated the code with requested changes

File: src/main/services/repack-tracker/helpers.ts (hunk 2 above)
Comment (JackEnx) [reply src/main/services/repack-tracker/helpers.ts:8]: It does indeed save the page. The reason behind is that we keep the page column in order to track and resume from the page that the scraper left at if it crashes or simply cannot move any forward.

For now I think it's best to leave the name as it is.

image

File: package.json [hunk 3]
```diff
@@ -53,6 +53,7 @@
     "jsdom": "^24.0.0",
     "lodash-es": "^4.17.21",
     "lottie-react": "^2.4.0",
+    "lucide-react": "^0.378.0",
```
Comment (hydralauncher) [inline package.json:56]: Do you really need a whole library for two icons? Can't you use the svgs individiually instead?

File: src/locales/en/translation.json [hunk 4]
```diff
@@ -96,7 +96,20 @@
     "dont_show_it_again": "Don't show it again",
     "copy_to_clipboard": "Copy",
     "copied_to_clipboard": "Copied",
-    "got_it": "Got it"
+    "got_it": "Got it",
+    "repack_language_code": {
```
Comment (hydralauncher) [inline src/locales/en/translation.json:100]: I don't think this is really necessary. Plase check this out:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/DisplayNames
(This applies for all the translations)

File: src/main/events/index.ts [hunk 5]
```diff
@@ -27,6 +27,7 @@ import "./torrenting/start-game-download";
 import "./user-preferences/get-user-preferences";
 import "./user-preferences/update-user-preferences";
 import "./user-preferences/auto-launch";
+import "./repacks/magnet-data"
```
Comment (hydralauncher) [inline src/main/events/index.ts:30]: Please consider this as being part of the catalogue instead

File: src/main/services/open-web-torrent.ts [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+      error: {
+        code: number;
+        message: string;
+      };
+    };
+
+const httpsAgent = new Agent({
+  rejectUnauthorized: false,
+});
+const axiosRef = axios.create({
```
Comment (hydralauncher) [inline src/main/services/open-web-torrent.ts:19]: You don't need a whole axios instance for a single request;
Plus, you should use https://www.npmjs.com/package/webtorrent-health instead (it's the same library used by the checker).

File: src/main/services/repack-tracker/1337x.ts [hunk 7]
```diff
@@ -3,7 +3,7 @@ import { JSDOM } from "jsdom";
 import { formatUploadDate } from "@main/helpers";
 
 import { Repack } from "@main/entity";
-import { requestWebPage, savePage } from "./helpers";
+import { requestWebPage, saveRepacks } from "./helpers";
```
Comment (hydralauncher) [inline src/main/services/repack-tracker/1337x.ts:6]: Please keep `savePage` instead

File: src/preload/index.ts [hunk 8]
```diff
@@ -112,4 +112,5 @@ contextBridge.exposeInMainWorld("electron", {
   showOpenDialog: (options: Electron.OpenDialogOptions) =>
     ipcRenderer.invoke("showOpenDialog", options),
   platform: process.platform,
+  getMagnetData: (magnet: string) => ipcRenderer.invoke("getMagnetData", magnet),
```
Comment (hydralauncher) [inline src/preload/index.ts:115]: I think it's worth changing the event name to `getMagnetHealth` or something

File: src/renderer/src/components/tag/tag.css.ts [hunk 9]
```diff
@@ -0,0 +1,11 @@
+import { style } from "@vanilla-extract/css";
+
+export const tagStyle = style({
```
Comment (hydralauncher) [inline src/renderer/src/components/tag/tag.css.ts:3]: No formatting on this file

File: src/renderer/src/components/tag/tag.css.ts [hunk 10]
```diff
@@ -0,0 +1,11 @@
...
+  borderRadius: '3px',
+  border: '1px solid #FFFFFF33',
```
Comment (hydralauncher) [inline src/renderer/src/components/tag/tag.css.ts:5]: Use the border color from the theme vars please

File: src/renderer/src/components/tag/tag.css.ts [hunk 11]
```diff
@@ -0,0 +1,11 @@
...
+  padding: '2px 5px',
```
Comment (hydralauncher) [inline src/renderer/src/components/tag/tag.css.ts:6]: Don't use arbitrary spacing, use `SPACING_UNIT` from the vars instead

File: src/renderer/src/pages/game-details/seeders-and-peers/useMagnetData.tsx [hunk 12]
```diff
@@ -0,0 +1,78 @@
+import { useEffect, useState } from "react";
```
Comment (hydralauncher) [inline src/renderer/src/pages/game-details/seeders-and-peers/useMagnetData.tsx:1]: Do you really think that it's necessary to create a hook that is going to be used in a single place?

**Review by hydralauncher — CHANGES_REQUESTED** [review]

File: src/main/services/open-web-torrent.ts (hunk 6 above)
Comment (ChristoferMendes) [reply src/main/services/open-web-torrent.ts:19]: I didn't know about that, thanks!

File: src/renderer/src/pages/game-details/seeders-and-peers/useMagnetData.tsx (hunk 12 above)
Comment (ChristoferMendes) [reply src/renderer/src/pages/game-details/seeders-and-peers/useMagnetData.tsx:1]: I prefer to organize components with custom hooks and utilities to keep them focused on rendering, making the code more modular and maintainable. What do you think?

File: src/main/services/repack-tracker/helpers.ts (hunk 2 above)
Comment (ChristoferMendes) [reply src/main/services/repack-tracker/helpers.ts]: Thank you for the information, reverting

File: package.json (hunk 3 above)
Comment (ChristoferMendes) [reply package.json]: Done

File: src/locales/en/translation.json (hunk 4 above)
Comment (ChristoferMendes) [reply src/locales/en/translation.json]: Perfect, thanks. Done

**Comment by zamitto** [comment]
@ChristoferMendes , just a heads up, there are lint errors to fix as well

**Comment by zamitto** [comment]
Repacks from onlinefix should have the multiplayer tag. Could you check this feature? It does not seem to be working here

image

**Comment by ChristoferMendes** [comment]
> Repacks from onlinefix should have the multiplayer tag. Could you check this feature? It does not seem to be working here
> 
> image

Will check

**Comment by ChristoferMendes** [comment]
> Repacks from onlinefix should have the multiplayer tag. Could you check this feature? It does not seem to be working here
> 
> image

Fixed! I was not checking for the repacker itself, only title

**Comment by zamitto** [comment]
You last thing I think.
Are all magnets returning 0 seeders and 0 peers for you as well?
image

**Comment by ChristoferMendes** [comment]
> You last thing I think.
> Are all magnets returning 0 seeders and 0 peers for you as well?
> image
> 

Strange, for me it was okay. I'll check it out

**Comment by zamitto** [comment]
> > You last thing I think.
> > Are all magnets returning 0 seeders and 0 peers for you as well?
> > image
> 
> Strange, for me it was okay. I'll check it out

Yeah, the log says timeout for all trackers. It's strange.

**Comment by JackEnx** [comment]
earlier I was checking this timeout issue, one way I did it was to check when empty, then i triggering a second request with a longer time, below is an example, its just a example

image

**Comment by ChristoferMendes** [comment]
> earlier I was checking this timeout issue, one way I did it was to check when empty, then i triggering a second request with a longer time, below is an example, its just a example
> 
> image
> 
> first request, 1 second
> 
> image
> 
> second request, 10 seconds after the first request
> 
> image

Applied retry logic in here: https://github.com/hydralauncher/hydra/pull/270/commits/7a2b693397ef051ec50aee0a8428d824b11a567e

File: src/main/services/web-torrent-data.ts [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+            return resolve(
+              webTorrentData.getSeedersAndPeers(magnet, retry + 1, timeout * 2)
+            );
+          }
+
+          return resolve({ peers, seeders: seeds });
+        }
+      );
+    });
+  },
```
Comment (zamitto) [inline src/main/services/web-torrent-data.ts:31]: i don't like the recursive function

could you replace with a loop solution?
maybe something like this (just an example):

```
  async getSeedersAndPeers(magnet: string) {
    let peers = 0;
    let seeds = 0;
    let retry = 0;

    while (retry < 3 && (!peers || !seeds)) {
      try {
        const data = await webTorrentHealth(magnet, {
          timeout: 1500 * (retry + 1),
        });

        peers = data.peers;
        seeds = data.seeds;
      } catch (e) {
        retry++;
      }
    }

    return { peers, seeders: seeds };
  },
```

**Review by zamitto — CHANGES_REQUESTED** [review]

File: src/main/services/web-torrent-data.ts (hunk 13 above)
Comment (ChristoferMendes) [reply src/main/services/web-torrent-data.ts]: Done

**Comment by ChristoferMendes** [comment]
> will try to finish reviewing your pr todat, we are kinda busy this week

Thank you!

**Review by zamitto — APPROVED** [review]

**Comment by zamitto** [comment]
@ChristoferMendes, I had merged your PR earlier, but we will actually  test it a little more before an actual merge

**Comment by zamitto** [comment]
If you could recreate the PR that would be nice

**Comment by ChristoferMendes** [comment]
> If you could recreate the PR that would be nice

Recreate it? Without changes?

**Pull request merged**